Use custom SSLSocketFactory to enable Server Name Indication for HTTPS connections

Crazyphil opened the pull request:
This is the pull request to solve issue #164.

william-ferguson-au commented:
I'm gearing up to release 4.6.0, so if you want this included you need to address the comments above.

Crazyphil commented:
Thanks for your comments. I'm currently rather busy, but I will see to look into your suggestions this weekend.

Crazyphil commented on src/main/java/org/acra/util/TlsSniSocketFactory.java:
```diff
@@ -0,0 +1,94 @@
+import javax.net.ssl.SSLSession;
+import javax.net.ssl.SSLSocket;
+
+@TargetApi(Build.VERSION_CODES.JELLY_BEAN_MR1)
```
SSLCertificateSocketFactory.setHostname() requires at least Jelly Bean, else a fallback is used. I've moved the annotation to the method that is affected by this.

Crazyphil commented:
I've modified the pull request according to your comments. Also, I've looked through the implementation of the SNI-aware socket factory in DAVdroid and added some of the changes there: https://github.com/bitfireAT/davdroid/blob/master/app/src/main/java/at/bitfire/davdroid/webdav/TlsSniSocketFactory.java

To use further features like SNI-aware SSL connections through proxies, ACRA would have to switch to the Apache HttpClient library instead of the Android one. This is not really necessary, however.

william-ferguson-au commented:
@Crazyphil I can't merge this as is as there are conflicts with #141 .
Can you please 
1) Take a copy of your changes
2) rebase to ACRA master
3) create a branch for your changes (you should always create a branch for changes),
4) Apply your change and merge them in
4) submit a new pull request from that branch.

william-ferguson-au commented:
Closed in favour of #221
